### PR TITLE
[depends: #1330] Interpreter fixes

### PR DESCRIPTION
**Pull request by tautschnig**
None of this is my work - these are commits already merged into develop (test-gen-support).

### CODE REVIEW

**Comment by tautschnig** [comment]
FWIW I just added as reviewers all those that have at least one commit in this series.

**Comment by tautschnig** [comment]
I'm looking at the linter failures and will submit a pull request to develop to address those first.

File: src/goto-programs/interpreter_class.h [hunk 1]
```diff
@@ -15,49 +15,188 @@ Author: Daniel Kroening, kroening@kroening.com
...
-    irep_idt identifier;
-    unsigned offset;
+    memory_cellt() :
+    value(0),
+    initialized(0)
+    {}
     mp_integer value;
+    // Initialized is annotated even during reads
+    // Set to 1 when written-before-read, -1 when read-before-written
+    mutable char initialized;
```
Comment (peterschrammel) [inline src/goto-programs/interpreter_class.h]: Probably, this should be turned into a three-valued enum.

File: src/goto-programs/interpreter.cpp [hunk 2]
```diff
@@ -121,147 +272,465 @@ void interpretert::step()
...
 }
 
 void interpretert::execute_decl()
 {
-  assert(PC->code.get_statement()==ID_decl);
+  assert(pc->code.get_statement()==ID_decl);
 }
 
+/// retrieves the member at offset
+/// \par parameters: an object and a memory offset
```
Comment (peterschrammel) [inline src/goto-programs/interpreter.cpp]: use \param

File: src/goto-programs/interpreter_evaluate.cpp [hunk 3]
```diff
@@ -13,39 +13,340 @@ Author: Daniel Kroening, kroening@kroening.com
...
+    dest[i]=value;
+  }
+}
+
+void interpretert::read_unbounded(
+  mp_integer address,
+  mp_vectort &dest) const
+{
+  // copy memory region
+  std::size_t address_val=integer2size_t(address);
```
Comment (mgudemann) [inline src/goto-programs/interpreter_evaluate.cpp]: could be `const`, too

File: src/goto-programs/interpreter_evaluate.cpp [hunk 4]
```diff
@@ -13,39 +13,340 @@ Author: Daniel Kroening, kroening@kroening.com
...
+    else
+      value=0;
+
+    dest.push_back(value);
+  }
+}
+
+/// reserves memory block of size at address
+void interpretert::allocate(
+  mp_integer address,
```
Comment (mgudemann) [inline src/goto-programs/interpreter_evaluate.cpp]: `const` reference ?

File: src/goto-programs/interpreter.cpp (hunk 2 above)
Comment (tautschnig) [reply src/goto-programs/interpreter.cpp]: `git grep -w '\\par' | wc -l` yields 273 -- I'll start fixing this, but it looks like a somewhat larger documentation cleanup.

File: src/goto-programs/interpreter_class.h (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/interpreter_class.h]: This is now #1289.

File: src/goto-programs/interpreter_evaluate.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-programs/interpreter_evaluate.cpp]: I have added this to #1289.

File: src/goto-programs/interpreter_evaluate.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-programs/interpreter_evaluate.cpp]: I have added this to #1289.

**Comment by smowton** [comment]
Happy with the smaller commits (perhaps consider folding the final bugfix commit into the appropriate sparse-vector commit?), the first one is too big to review meaningfully though. It looks like it is itself a big merge -- any chance of finding the original commits that constituted it?

**Comment by smowton** [comment]
Dismissed stale review request

**Comment by tautschnig** [comment]
I believe we are not going ahead with a selective develop->master transition, thus closing.

**Pull request closed without merging**